File: src/main/java/algo/twoPointers/BoatsToSavePeople_881.java
package algo.twoPointers;

import java.util.Arrays;

public class BoatsToSavePeople_881 {
  public static int numRescueBoats(int[] people, int limit) {
    int boats = 0;
    int len = people.length;
    Arrays.sort(people);

    int start = 0, end = len - 1;
    while (start <= end) {
      if (start == end) {
        boats++;
        break;
      }
      if (people[start] + people[end] <= limit) {
        boats++;
        start++;
        end--;
      } else {
        boats++;
        end--;
      }
    }

    return boats;
  }

  public static void main(String[] args) {
    System.out.println(numRescueBoats(new int[]{1,2}, 3));
    System.out.println(numRescueBoats(new int[]{3,2,2,1}, 3));
    System.out.println(numRescueBoats(new int[]{3,5,3,4}, 5));
    System.out.println(numRescueBoats(new int[]{5,1,4,2}, 6));
    System.out.println(numRescueBoats(new int[]{3,2,3,2,2}, 6));
  }
}

File: src/main/kotlin/algo/twoPointers/BoatsToSavePeople_881.kt
package algo.twoPointers

class KtBoatsToSavePeople_881 {
    fun numRescueBoats(people: IntArray, limit: Int): Int {
        var left = 0; var right = people.size - 1
        var boats = 0
        people.sort()
        while (left <= right) {
            if (left == right) {
                boats++
                break
            }
            if (people[left] + people[right] <= limit) {
                boats++
                left++
                right--
            } else {
                boats++
                right--
            }
        }
        return boats
    }
}

fun main() {
    val boat = KtBoatsToSavePeople_881()
    println(boat.numRescueBoats(intArrayOf(3,2,2,1), 3))
    println(boat.numRescueBoats(intArrayOf(3,5,3,4), 5))
    println(boat.numRescueBoats(intArrayOf(3,8,7,1,4), 9))
}